middle_end/flambda2/simplify_shared/closure_offsets.ml

```diff
@@ -689,145 +734,83 @@ module Greedy = struct
-      used_closure_vars offsets
-
-  let imported_and_used_offsets ~used_closure_ids ~used_closure_vars state =
-    match
-      (used_closure_ids, used_closure_vars : _ Or_unknown.t * _ Or_unknown.t)
-    with
-    | Known used_closure_ids, Known used_closure_vars ->
+  let imported_and_used_offsets ~used_names state =
+    match (used_names : _ Or_unknown.t) with
+    | Known
+        { closure_ids_normal;
+          closure_ids_in_types;
+          closure_vars_normal;
+          closure_vars_in_types
+        } ->
       state.used_offsets
-      |> collect_used_closure_ids state ~used_closure_ids
-      |> collect_used_closure_vars state ~used_closure_vars
-    | Unknown, Known _ | Known _, Unknown | Unknown, Unknown ->
-      EO.imported_offsets ()
-
-  (* Currently, it happens that some closure_ids/closure_vars can occur in
-     projections (and thus in the used_closure_ids/vars), but never occur in the
-     creation of a set of closures (and consequently have no slot, and thus are
-     not assigned any offset).
-
-     There are two main reasons why this could happen:
-
-     - we could have missed some sets of closures creations on the way up and
-     not recorded the associated constraints
-
-     - the closure_id/closure_var occur in dead code that was kept for some
-     reason (mainly due to the over-approximations made by data_flow, which
-     keeps alive more things than necessary). This regularly occurs when phantom
-     lets are generated, but can also occur without phantom let bindings.
-
-     Upon a missing offset for a closure_id/var, to_cmm will generate cmm
-     instructions that produce a segfault. This is safe in the second case, when
-     the closure_id/var with the missing offset occurs in dead code, but it is
-     incorrect in the first case.
-
-     The following check is intended to catch the first of these two cases, but
-     it cannot distinguish between the two cases, and this check results in a
-     lot of false positives. Thus we cannot always run this check. *)
-  let check_used_offsets state ~used_closure_ids ~used_closure_vars offsets =
-    if !Clflags.flambda_invariant_checks
-    then
-      match
-        (used_closure_ids, used_closure_vars : _ Or_unknown.t * _ Or_unknown.t)
-      with
-      | Known used_closure_ids, Known used_closure_vars ->
+      |> collect_used_closure_ids closure_ids_normal
+      |> collect_used_closure_ids closure_ids_in_types
+      |> collect_used_closure_vars closure_vars_normal
+      |> collect_used_closure_vars closure_vars_in_types
+    | Unknown -> EO.imported_offsets ()
+
+  (* We only want to keep closure vars that appear in the creation of a set of
+     closures, *and* appear as projection (at normal name mode). And we need to
+     mark closure_vars/ids that are not alive, as dead in the exported_offsets,
+     so that later compilation unit do not mistake that for a missing offset
+     info on a closure_var/id. *)
```

Comment:
You mentioned on Slack that it's possible for us to have a dead closure id that gets used in projections; seems worth noting here how that could happen.